docs/src/references/hsm-support.md

```diff
@@ -1,242 +1,228 @@
+   user from the previous step.
+
+    ```sh
+    PUB_PRIV_KEY=$(
+        cat "$(tedge config get device.key_path)" && cat "$(tedge config get device.cert_path)"
+    )
+    softhsm2-util \
+        --import <(echo "$PUB_PRIV_KEY") \
+        --token my-token \
+        --label my-key \
+        --id 01 \
+        --pin 123456 \
+    ```
+
+4. Get the URI of the key
+
+    First, see what tokens are available
+
+    ```sh
+    p11tool --list-tokens
+    ```
+
+    ```sh title="p11tool --list-tokens"
+    ...
+    Token 2:
+        URL: pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=83f9cf49039c051a;token=my-token
+        Label: my-token
+        Type: Generic token
+        Flags: RNG, Requires login
+        Manufacturer: SoftHSM project
+        Model: SoftHSM v2
+        Serial: 83f9cf49039c051a
+        Module: /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so
+    ...
+    ```
+
+    Now check if the private key object is in the token. You may need to login, provide the regular
+    user PIN and also provide token URL(URI) if multiple tokens are connected:
+
+    ```sh
+    sudo p11tool --login --set-pin=123456 --list-privkeys "pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=83f9cf49039c051a;token=my-token"
+    ```
+    ```sh title="sudo p11tool --login --set-pin=123456 --list-privkeys "pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=83f9cf49039c051a;token=my-token""
+    Object 0:
+        URL: pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=83f9cf49039c051a;token=my-token;id=%01;object=my-key;type=private
+        Type: Private key (EC/ECDSA-SECP256R1)
+        Label: my-key
+        Flags: CKA_PRIVATE; CKA_SENSITIVE; 
+        ID: 01
+    ```
+
+### Part 2: thin-edge setup
+
+Next, we're going to configure `tedge` to use the token directly using module mode.
+
+If that mode doesn't work for you, because of the token can't be accessed for some reason or you
+can't dynamically load the PKCS #11 library, see how to [use `tedge-p11-server`](./tedge-p11-server.md).
+<!-- TODO: put instructions for tp11s in its dedicated page -->
+
+Using the module mode, the cryptoki module will be loaded by the dynamic loader
+and used for signing. If there are many tokens or private keys we also need to
+provide the URI for the key to select a correct one.
+
+5. Enable the module mode and set the module and the key URI.
 ```sh
-echo "module: /usr/lib/x86_64-linux-gnu/pkcs11/opensc-pkcs11.so" > /usr/share/p11-kit/modules/opensc-pkcs11.module
-echo "module: /usr/lib/x86_64-linux-gnu/pkcs11/p11-kit-client.so" > /etc/pkcs11/modules/p11-kit-client.module
+tedge config set device.cryptoki.mode module
+tedge config set device.cryptoki.module_path /usr/lib/x86_64-linux-gnu/softhsm/libsofthsm2.so
+# optional if there are many tokens/keys
+tedge config set device.key_uri "pkcs11:model=SoftHSM%20v2;manufacturer=SoftHSM%20project;serial=83f9cf49039c051a;token=my-token;id=%01;object=my-key;type=private"
```

Comment:
Is this correct? Shouldn't this be the contrary i.e that `device.key_uri` is optional when there is *only one* key?

Reply:
comment removed in 8fc4a62b3e

Decided that since we talk about key selection in its section, it's better to mention it there than add comments here.
And indeed the comment was wrong, the setting is effectively optional only if there is 1 token and 1 key.